databricks_cli/stack/api.py

```diff
@@ -0,0 +1,353 @@
+                             " stack deployment" % job_name)
+        elif len(jobs_same_name) == 1:
+            existing_job = jobs_same_name[0]
+            creator_name = existing_job['creator_user_name']
+            timestamp = existing_job['created_time'] / MS_SEC  # Convert to readable date.
+            date_created = datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
+            click.echo("Warning: Job exists with same name '%s' created by %s on %s. Job will "
+                       "be overwritten" % (job_name, creator_name, date_created))
+            self.update_job(job_settings, existing_job['job_id'])
+            return existing_job['job_id']
+        else:
+            click.echo("Creating new job")
+            job_id = self.jobs_client.create_job(job_settings)['job_id']
+            return job_id
+
+    def update_job(self, job_settings, job_id):
+        """
+        Given job settings and an existing job_id of a job, update the job settings on databricks.
+
+        :param job_settings: job settings to update the job with.
+        :param job_id: physical job_id of job in databricks server.
+        """
+        click.echo("Updating Job")
+        self.jobs_client.reset_job({'job_id': job_id, 'new_settings': job_settings})
+
+    def deploy_job(self, resource_id, resource_properties, physical_id=None):
+        """
+        Deploys a job resource by either creating a job if the job isn't kept track of through
+        the physical_id of the job or updating an existing job. The job is created or updated using
+        the the settings specified in the inputted job_settings.
+
+        :param resource_id: The stack-internal resource ID of the job.
+        :param resource_properties: A dict of the Databricks JobSettings data structure
+        :param physical_id: A dict object containing 'job_id' field of job identifier in Databricks
+        server
+
+        :return: tuple of (physical_id, deploy_output), where physical_id contains a 'job_id' field
+        of the physical job_id of the job on databricks. deploy_output is the output of the job
+        from databricks when a GET request is called for it.
+        """
+        job_settings = resource_properties  # resource_properties of jobs are solely job settings.
+        click.echo("Deploying job '%s' with settings: \n%s \n" % (resource_id, json.dumps(
+            job_settings, indent=2, separators=(',', ': '))), nl=False)
+
+        if physical_id and 'job_id' in physical_id:
+            job_id = physical_id['job_id']
+            try:
+                self.update_job(job_settings, physical_id['job_id'])
+            except HTTPError:
+                # If updating a job fails, create the job with put_job
+                job_id = self.put_job(job_settings)
+        else:
+            job_id = self.put_job(job_settings)
+
+        job_url = "%s#job/%s" % (self.host, str(job_id))
```

Comment:
@xuanwang14 Should I leave URL generation out of this PR?

Reply:
I feel the value of URL is not very clear considering that the logic here maybe hard to maintain (and not reliable). That being said, I do not have a strong opinion here.